src/term/mod.rs

```diff
@@ -398,150 +399,275 @@ impl Term {
     }
   }
 
-  /* Common checks and transformations */
-
-  /// Substitute the occurrences of a variable in a term with the given term.
-  /// Caution: can cause invalid shadowing of variables if used incorrectly.
-  /// Ex: Using subst to beta-reduce (@a @b a b) converting it into (@b b).
-  pub fn subst(&mut self, from: &Name, to: &Term) {
+  /* Iterators */
+  pub fn children(&self) -> impl DoubleEndedIterator<Item = &Term> {
     match self {
-      Term::Lam { nam: Some(nam), .. } if nam == from => (),
-      Term::Lam { bod, .. } => bod.subst(from, to),
-      Term::Var { nam } if nam == from => *self = to.clone(),
-      Term::Var { .. } => (),
-      // Only substitute scoped variables.
-      Term::Chn { bod, .. } => bod.subst(from, to),
-      Term::Lnk { .. } => (),
-      Term::Let { pat, val, nxt } => {
-        val.subst(from, to);
-        if !pat.binds().contains(from) {
-          nxt.subst(from, to);
-        }
+      Term::Mat { args, rules } => {
+        ChildrenIter::Many(Box::new(args.iter().chain(rules.iter().map(|r| &r.body))))
```

Comment:
Boxing these iterators is unnecessary overhead, and the `dyn` will prevent rust from optimizing the iteration. Is there a reason you can't use the macro I wrote, with the generics? It would let you avoid boxing these iterators.

Reply:
I don't like how the macro approach created a bunch of opaque structures.
It was hard to work with as I tried to extend them to do more things (which I ultimately failed to do in a simple way) and I got errors from things created in the macro.

I figured since performance is not a huge concern for hvm-lang right now I'd leave it like this which will probably be easier for me to try to improve in the future.